File: src/main/java/cadenza/NeutralException.java
package cadenza;

import cadenza.values.Neutral;
import com.oracle.truffle.api.nodes.ControlFlowException;

public class NeutralException extends ControlFlowException {
  private static final long serialVersionUID = 5587798688299594259L;
  public final Neutral term;
  public NeutralException(Neutral term) { this.term = term; }
}

File: src/cadenza/data/NeutralException.kt
package cadenza.data

import cadenza.semantics.Type
import com.oracle.truffle.api.nodes.SlowPathException

class NeutralException(val type: Type, val term: Neutral) : SlowPathException() {
  fun get() = NeutralValue(type, term)

  fun apply(rands: Array<out Any?>): Nothing = neutral(
    rands.indices.fold(type) { currentType, _ -> (currentType as Type.Arr).result },
    term.apply(rands)
  )

  companion object { const val serialVersionUID : Long = 1L }
}

@Throws(NeutralException::class)
fun neutral(type: Type, term: Neutral) : Nothing = throw NeutralException(type, term)

fun NeutralValue.raise() : Nothing = throw NeutralException(type, term)

fun <T> throwIfNeutralValue(x: T): T = if (x !is NeutralValue) x else x.raise()